examples/grovers.rs

```diff
@@ -107,33 +106,101 @@ fn grovers_algorithm(winner: usize, regsize: usize) -> GroversResult {
 // other states.
 fn diffuser(reg: &mut Register) {
     reg.apply_all(&hadamard(0));
-    reg.apply_all(&not(0));
-    reg.apply(&cz(&(0..reg.size() - 1).collect(), reg.size() - 1));
-    reg.apply_all(&not(0));
+    reg.apply(&diffusion_operation(reg.size()));
     reg.apply_all(&hadamard(0));
 }
 
+/// The oracle function U_f reflects the "winner" state
+///
+/// ***Panics*** if the Operation fails to be created
 pub fn oracle_operation(regsize: usize, winner: usize) -> Operation {
     let n: usize = 2_usize.pow(regsize as u32);
     let mut matrix: Array2<f64> = Array2::<f64>::zeros((n, n));
     for i in 0..n {
         matrix.row_mut(i)[i] = if i == winner { -1.0 } else { 1.0 };
     }
 
-    let op = Operation::new(real_arr_to_complex(matrix), (0..regsize).collect()).expect("Could not create oracle operation");
+    let op = Operation::new(real_arr_to_complex(matrix), (0..regsize).collect())
+        .expect("Could not create oracle operation");
     op
 }
 
-// Calculates the optimal number of iterations needed for U_sU_f
-// to get good amplitudes for the target states.
-fn iterations(search_space: usize) -> usize {
-    (PI / 4.0 * 2.0_f64.pow(search_space as f64).sqrt() / 1.0).floor() as usize
+/// Creates the diffuser operation U_s
+///
+/// ***Panics*** if the Operation fails to be created
+pub fn diffusion_operation(regsize: usize) -> Operation {
```

Comment:
Hard to understand what it does. A few comments should help.

Reply:
Should be easier now